### PR TITLE
PEFT LLM training example 

### PR DESCRIPTION
**Pull request by GentleZhu**
*Issue #, if available:*

*Description of changes:*
Added PEFT LLM training using customized model interface

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: examples/peft_llm_gnn/llm_gnn_model.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+                 target_ntype,     # the node type to be predict
+                 use_norm = True,   # use normalization or not, default is True
+                 alpha_l2norm = 0,
+                 lr = 0.001):
+        super(LLMGraphModel, self).__init__()
+        self.num_layers = num_layers
+        self.target_ntype=target_ntype
+        self.alpha_l2norm = alpha_l2norm
+        self.lr = lr
+        model_id = "facebook/opt-2.7b"
```
Comment (zheng-da) [inline examples/peft_llm_gnn/llm_gnn_model.py:55]: this should be hard coded.

File: examples/peft_llm_gnn/llm_gnn_model.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        self.config = AutoConfig.from_pretrained(model_id, num_labels=out_dim)
+        base_model = AutoModelForSequenceClassification.from_pretrained(
+            model_id,config=self.config
+        )
+        peft_config = LoraConfig(task_type="SEQ_CLS", inference_mode=False, r=8, lora_alpha=16, lora_dropout=0.1)
```
Comment (zheng-da) [inline examples/peft_llm_gnn/llm_gnn_model.py:60]: we probably should also make lora hyperparameters configurable.

File: examples/peft_llm_gnn/llm_gnn_model.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        for lm_config in node_lm_configs:
+            # A list of node types sharing the same lm model
+            lm_ntypes = lm_config["node_types"]
+            lm_node_feats = get_lm_node_feats(g, lm_feat_names, lm_ntypes)
+            for ntype, feats in lm_node_feats.items():
+                assert ntype not in self._lm_node_feats, \
+                        f"More than one BERT model runs on Node {ntype}."
+                self._lm_node_feats[ntype] = feats
+        for _ in range(num_layers):
+            pass
```
Comment (zheng-da) [inline examples/peft_llm_gnn/llm_gnn_model.py:73]: what does this do?

File: examples/peft_llm_gnn/llm_gnn_model.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+    def forward(self, blocks, node_feats, edge_feats, labels, input_nodes):
+        # TODO (qzhuamzn): use GNNs to generate graph tokens
+        h = {}
+        output_nodes = blocks[-1].dstdata["_ID"]
+
+        input_ids = self._lm_node_feats[self.target_ntype]["input_ids"][output_nodes].to(self.llm.device)
```
Comment (zheng-da) [inline examples/peft_llm_gnn/llm_gnn_model.py:80]: it's better not to hard code "input_ids"

File: examples/peft_llm_gnn/llm_gnn_model.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+        attention_mask = self._lm_node_feats[self.target_ntype]["attention_mask"][output_nodes].to(self.llm.device)
+        # TODO (qzhuamzn): modify input_ids into input_embeds=[graph_tokens, input_embeds] to support GPEFT
+        model_output = self.llm(input_ids=input_ids, attention_mask=attention_mask, labels=labels[self.target_ntype])
+        return model_output.loss
```
Comment (zheng-da) [inline examples/peft_llm_gnn/llm_gnn_model.py:84]: this is a general peft model? how is it configured for the node task?

**Comment by zheng-da** [comment]
this PR is to use the regular PEFT for a node prediction task?

File: examples/peft_llm_gnn/llm_gnn_model.py (hunk 2 above)
Comment (GentleZhu) [reply examples/peft_llm_gnn/llm_gnn_model.py:60]: I plan to add configurable lora next PR with GNN encoded tokens. What do you think?

File: examples/peft_llm_gnn/llm_gnn_model.py (hunk 3 above)
Comment (GentleZhu) [reply examples/peft_llm_gnn/llm_gnn_model.py:73]: leave for gnn encoded tokens

File: examples/peft_llm_gnn/llm_gnn_model.py (hunk 5 above)
Comment (GentleZhu) [reply examples/peft_llm_gnn/llm_gnn_model.py:84]: configured via LoraConfig

File: python/graphstorm/eval/utils.py [hunk 6]
```diff
@@ -304,7 +304,7 @@ def broadcast_data(rank, world_size, data_tensor):
     if get_backend() == "gloo":
         device = "cpu"
     elif get_backend() == "nccl":
-        assert data_tensor.is_cuda
```
Comment (GentleZhu) [inline python/graphstorm/eval/utils.py:307]: @zheng-da Do you have concerns about this change? I found data tensor is not on gpu in nccl

File: examples/peft_llm_gnn/llm_gnn_model.py (hunk 4 above)
Comment (GentleZhu) [reply examples/peft_llm_gnn/llm_gnn_model.py:80]: How do we do this in other LM models?

File: examples/peft_llm_gnn/llm_gnn_model.py (hunk 5 above)
Comment (classicsong) [reply examples/peft_llm_gnn/llm_gnn_model.py:84]: So the loss is from AutoModelForSequenceClassification?

File: examples/peft_llm_gnn/llm_gnn_model.py [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+
+        return torch.optim.Adam(self.parameters(), lr=self.lr)
+    
+    def restore_model(self, restore_model_path):
+        self.llm = AutoModelForSequenceClassification.from_pretrained(restore_model_path, config=self.config)
+
+    def save_model(self, model_path):
+        os.makedirs(model_path, exist_ok=True)
+        os.chmod(model_path, 0o767)
+        if torch.distributed.get_rank() == 0:
```
Comment (classicsong) [inline examples/peft_llm_gnn/llm_gnn_model.py:113]: You should use graphstorm get_rank()

**Review by classicsong — COMMENTED** [review]
Basically, in This PR, we do not have GNN in the model,right?

File: examples/peft_llm_gnn/llm_gnn_model.py (hunk 3 above)
Comment (zheng-da) [reply examples/peft_llm_gnn/llm_gnn_model.py:73]: you should add comments. but i think it's better to just add comments here instead of having the code like this.

File: examples/peft_llm_gnn/llm_gnn_model.py (hunk 4 above)
Comment (zheng-da) [reply examples/peft_llm_gnn/llm_gnn_model.py:80]: i think GraphStorm uses the constant `TOKEN_IDX` for "input_ids".

File: python/graphstorm/eval/utils.py (hunk 6 above)
Comment (zheng-da) [reply python/graphstorm/eval/utils.py:307]: the question is which GPU is used here?

File: examples/peft_llm_gnn/gen_graph.sh [hunk 8]
```diff
@@ -0,0 +1,9 @@
+WORKSPACE=/shared_data
+dataset=amazon_review
+domain=$1
+rm -rf ./data/"$dataset"/predict_pt/"$domain"
+python "$WORKSPACE"/graphstorm/tools/partition_graph.py --dataset graph_200_200_nc.opt_gs.bin \
+    --filepath "$WORKSPACE"/GraphPEFT/dataset/"$dataset"/"$domain" --target-ntype item \
+    --nlabel-field item:label --num-parts 1\
+    --output ./data/"$dataset"/predict_pt/"$domain"
```
Comment (zheng-da) [inline examples/peft_llm_gnn/gen_graph.sh:8]: use gconstruct to construct the graph.

File: python/graphstorm/eval/utils.py (hunk 6 above)
Comment (GentleZhu) [reply python/graphstorm/eval/utils.py:307]: I think .cuda goes to the correpsonding rank of the device.

File: examples/peft_llm_gnn/llm_gnn_model.py (hunk 5 above)
Comment (GentleZhu) [reply examples/peft_llm_gnn/llm_gnn_model.py:84]: Changed into gsmodel.ClassifyLossFunc

File: python/graphstorm/eval/utils.py [hunk 9]
```diff
@@ -304,7 +304,7 @@ def broadcast_data(rank, world_size, data_tensor):
...
+        data_tensor = data_tensor.cuda()
```
[the remark this answers is not in the document]
Comment (GentleZhu) [reply python/graphstorm/eval/utils.py:307]: I think so.

File: examples/peft_llm_gnn/README.md [hunk 10]
```diff
@@ -0,0 +1,51 @@
+This folder contains the data processing script to process the raw Amazon Review dataset
+downloaded from https://cseweb.ucsd.edu/~jmcauley/datasets/amazon_v2/. We  domain Video
```
Comment (zheng-da) [inline examples/peft_llm_gnn/README.md:2]:
```suggestion
downloaded from https://cseweb.ucsd.edu/~jmcauley/datasets/amazon_v2/. We use domain Video
```

File: examples/peft_llm_gnn/llm_gnn_model.py [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+                lm_feats[ntype][lm_fname] = g.nodes[ntype].data[lm_fname]
+
+    return lm_feats
+
+
+class LLMGraphModel(gsmodel.GSgnnNodeModelBase):
+    """ LLMGNN Model Class
+
+    Parameters
+    ----------
```
Comment (zheng-da) [inline examples/peft_llm_gnn/llm_gnn_model.py:65]: add docs on the arguments.

File: examples/peft_llm_gnn/llm_gnn_model.py [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+                 use_norm = True,   # use normalization or not, default is True
+                 alpha_l2norm = 0,
+                 lr = 0.001):
+        super(LLMGraphModel, self).__init__()
+        self.num_layers = num_layers
+        self.target_ntype=target_ntype
+        self.alpha_l2norm = alpha_l2norm
+        self.lr = lr
+        # assume only one LLM is used
+        model_id = node_lm_configs[0]['model_name']
```
Comment (zheng-da) [inline examples/peft_llm_gnn/llm_gnn_model.py:78]: because GPEFT only runs LLM on one node type?

Comment (zheng-da) [reply examples/peft_llm_gnn/llm_gnn_model.py:78]: if so, should we allow users to specify which node type and its corresponding LLM model?

File: examples/peft_llm_gnn/llm_gnn_model.py [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+                 target_ntype,     # the node type to be predict
+                 use_norm = True,   # use normalization or not, default is True
+                 alpha_l2norm = 0,
+                 lr = 0.001):
+        super(LLMGraphModel, self).__init__()
+        self.num_layers = num_layers
+        self.target_ntype=target_ntype
+        self.alpha_l2norm = alpha_l2norm
+        self.lr = lr
+        # assume only one LLM is used
```
Comment (zheng-da) [inline examples/peft_llm_gnn/llm_gnn_model.py:77]: I think it's better to add a comment to explain that only one LLM is supported. also, we should verify that a user only specify one LLM. if a user specifies more than one LLM, we should print a warning.

File: examples/peft_llm_gnn/llm_gnn_model.py [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+        logits = self.out(model_output.last_hidden_state[:,-1,:])
+        if return_proba:
+            return logits.argmax(dim=-1), torch.softmax(logits, 1)
+        else:
+            return logits.argmax(dim=-1), logits
+        
+    def create_optimizer(self):
+        # Here we assume there are no sparse embeddings.
+        pytorch_total_params = sum(p.numel() for p in self.parameters() if p.requires_grad)
+        print(f"Num of trainable params: {pytorch_total_params}")
```
Comment (zheng-da) [inline examples/peft_llm_gnn/llm_gnn_model.py:131]: use logging.debug

File: examples/peft_llm_gnn/llm_gnn_model.py [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+
+        return torch.optim.Adam(self.parameters(), lr=self.lr)
+    
+    def restore_model(self, restore_model_path):
```
Comment (zheng-da) [inline examples/peft_llm_gnn/llm_gnn_model.py:135]: if you implement restore_model, you need to implement save_model as well.

File: examples/peft_llm_gnn/preprocess_amazon_review.py [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+    return item_label, label_id
+
+
+
+def construct_graph(directory_path, ood_fields = ['Video_Games, Automotive']):
+    random.seed(2023)
+    
+    cnt = 0
+    for dirpath, dirnames, filenames in os.walk(directory_path):
+        print(filenames)
```
Comment (zheng-da) [inline examples/peft_llm_gnn/preprocess_amazon_review.py:123]: remove this print or print more meaningful message

File: examples/peft_llm_gnn/llm_gnn_model.py [hunk 17]
```diff
@@ -0,0 +1,163 @@
...
+        )
+        peft_config = LoraConfig(inference_mode=False, r=8, lora_alpha=16, lora_dropout=0.1)
+        self.llm = get_peft_model(base_model, peft_config)
+        lm_feat_names = [TOKEN_IDX, ATT_MASK_IDX]
+        self._lm_node_feats = {}
+        for lm_config in node_lm_configs:
+            # A list of node types sharing the same lm model
+            lm_ntypes = lm_config["node_types"]
+            if lm_config["model_name"] != model_id:
+                logging.warning("Mutiple LLM model name is found. Only one LLM is supported.")
```
Comment (zheng-da) [inline examples/peft_llm_gnn/llm_gnn_model.py:110]:
```suggestion
                logging.warning("Mutiple LLM model names are found in the config. However, the example only supports one LLM.")
```

**Review by zheng-da — APPROVED** [review]
